library/alloc/src/collections/btree/map/entry.rs

```diff
@@ -312,22 +313,39 @@ impl<'a, K: Ord, V> VacantEntry<'a, K, V> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn insert(self, value: V) -> &'a mut V {
-        let out_ptr = match self.handle.insert_recursing(self.key, value) {
-            (Fit(_), val_ptr) => {
-                // SAFETY: We have consumed self.handle and the handle returned.
-                let map = unsafe { self.dormant_map.awaken() };
-                map.length += 1;
-                val_ptr
-            }
-            (Split(ins), val_ptr) => {
-                drop(ins.left);
+        let out_ptr = match self.handle {
+            None => {
                 // SAFETY: We have consumed self.handle and the reference returned.
                 let map = unsafe { self.dormant_map.awaken() };
-                let root = map.root.as_mut().unwrap();
-                root.push_internal_level().push(ins.kv.0, ins.kv.1, ins.right);
+                let (root, out_ptr) = {
```

Comment:
I was not sure if I should make the `HandleRef::<Mut<'a>, K, V, LeafOrInternal>::cast_to_leaf_unchecked` public, or use `force` to convert a `LeafOrInternal` node to a `Leaf` node. So I did it everything on the new created `Leaf` node before it was `.forget()`ed.

As you see that, directly returning the `&'a mut V` requires borrowing from `map.root`, which is a `LeafOrInternal` node. I cannot borrow from the new created node before it is put into `map.root` since being borrowed and then moved is not allowed.

Reply:
I think you did well in explicit typing and avoiding type casts - I made `NodeRef::new_leaf` public in other PRs. I'm perhaps saying you didn't take it far enough: just move the `.forget_type()` to the assignment to `map.root`, and then the block seems completely superfluous to me (and to the compiler, when I try it).

I'm not quite sure what you mean with "directly returning", is it the fact that the `&mut V` is cast to a `*mut V` and back again? Indeed it's ugly here and even uglier elsewhere, and frankly I am not sure that there isn't undefined behaviour, now or in the making, even if Miri says there isn't, but anyway that's not at all what I meant with this comment.

Reply:
> I'm not quite sure what you mean with "directly returning"

Here is where I messed up the `*mut V`. The inner most `insert_fit` creates a `&mut V`,
https://github.com/rust-lang/rust/blob/3d1eaf4b6203ab0168da5b99049942385aa753e0/library/alloc/src/collections/btree/node.rs#L854
but it is immediately coerced to a `*mut V`, and is finally promoted to a `&mut V` again before returned to the user.